Feature clickable routes

jschleic opened the pull request:
First implementation using vector tiles for zoom levels > 12.
Please comment on the code.

Implements #29.

As next step I'll use this layer for segment highlighting as well (hovering in the route list).

lonvia commented:
Works nicely already and code looks good in general. Some random comments.

My main complained would be a lack of visual feedback, i.e. it's not at all clear where you can click. Can we change the mouse cursor when hovering over route lines? Or maybe ever so lightly highlight the line on mouseover? That latter solution would also allow to remove the render hack for for EachFeatureAtPixel which unfortunately is ever so slightly visible as an artifact.

If above is possible, then it is also okay to simply do nothing if no feature can be found under the cursor. I find it a bit confusing that an empty route list window is opened.
 
I'm also not sure about forwarding the bbox to the route list. I agree that it fits nicely into the existing API but it does put the burden of unnecessary geometry computations on the server. I would prefer to add another API endpoint 'by-ids?ids=123,43,234' and then hand the relation id list that the click event computes to the routelist URL. (In the long run, the vector tile should contain sufficient information about the relations, so that the route list can be displayed without involving the server at all. But I need to settle the vector tile format first.)

Note to self: must get vector tile API into production-worthy state asap.

jschleic commented:
Thanks for your feedback.

Opening an empty route window was not my intention. I've corrected that now.

An 'by-ids'-API is also good, I've implemented that also (see the amended commit a minute ago).

I don't know the details of OpenLayers, but I think: Hovering like in ol.interaction is easily possible with a Vector layer but we have ol.source.VectorTile source in an ol.layer.VectorTile layer. There the only option is to use forEachFeatureAtPixel at every mousemove event. But then the routes have to be visible, which is the case, after you have implemented the rendering.
So the 'render hack' is only necessary, until your browser-rendering is implemented and then we can try, if a cursor change is feasible on hovering.

lonvia commented:
I see what you mean. That makes it indeed difficult to have mouseover feedback. So let's drop the idea for now.

I still find the rendering unnerving though. So I've played a bit with OpenLayers and it looks like Vector and VectorTile can be used more or less exchangably. The resulting Vector-based implementation here: https://github.com/lonvia/waymarked-trails-site/tree/clickable-routes-with-vector I would suggest going with Vector for now and then switch to VectorTile as soon as we start actually rendering the ways.

I'll add a couple more nitpicking comments in the code.

lonvia commented on api/listings.py:
```diff
@@ -91,6 +91,25 @@ def by_area(self, bbox, limit=None, **params):
     @cherrypy.tools.json_out()
+    def by_ids(self, ids, limit=None, **params):
+        i = ids.split(',')
+        limit = self.num_param(limit, 20, 100)
```
The limit parameter does not make sense here as we know exactly know how many results will be returned. Cut the 'i' array at 100.

lonvia commented on api/listings.py:
```diff
@@ -91,6 +91,25 @@ def by_area(self, bbox, limit=None, **params):
+        h = mapdb.tables.hierarchy.data
+
+        res = sa.select([r.c.id, r.c.name, r.c.intnames, r.c.symbol, r.c.level])\
+               .where(r.c.id.in_(i))\
```
Make sure i only contains valid integers.

lonvia commented on api/listings.py:
```diff
@@ -91,6 +91,25 @@ def by_area(self, bbox, limit=None, **params):
+               .order_by(r.c.level, r.c.name)\
+               .limit(limit)
+
+        return self.create_list_output('ids', ",".join(i),
```
Just 'i' for second parameter. It will then be correctly rendered as a json array.

lonvia commented on frontend/static/js/app.js:
```diff
@@ -167,14 +167,30 @@ Osgende.RouteList = function(map, container) {
+    .on("refresh", function() {
+      var ids = decodeURI(window.location.hash.replace(
+        new RegExp("^(?:.*[&\\?]ids(?:\\=([^&]*))?)?.*$", "i"), "$1"));
+      if(ids) {
```
That 'if' looks unnecessary. update_list(ids) works in both cases.

lonvia commented on frontend/static/js/map.js:
```diff
@@ -110,6 +110,43 @@ Osgende.BaseMapControl = function(settings) {
+        if(rels)
+          relations = relations.concat(rels);
+      });
+      if(relations.length === 1)
```
General style comment: please leave a space after 'if', 'for' etc and adjust bracket style to rest of code (opening bracket on end of line).

jschleic commented:
Thanks for your comments. I've amended the commit and added your VectorLayer change.

lonvia commented:
Merged and deployed for the route layers. I've disabled the feature for the Slopes map for now, as that map needs a bit of special handling for the ids.

One thing to note: I had to tweak the list of relations a bit more as it was missing super relations. The list now contains all top-level relations that go over the segment, i.e. the same routes that are shown in the route list window. I think that this is what people most expect. However, it means that the vector layer tiles cannot be used to highlight route segments because subrelations with the same network level will be missing.

I've also moved the vector tile endpoint under /api/tiles. For Javascript there is a variable Osgende.VTILE_URL to allow for further urls changes in the future.

Thanks again for the implementation, that's an awesome feature to have.
